### ISSUE TITLE
Users need to be able to install custom persistent packages

### ISSUE THREAD

**Issue by moorepants** [LibreTexts/metalc#146]
I don't think there is an open issue about this and I know we have these instructions https://github.com/LibreTexts/default-env#creating-your-own-custom-conda-environment, but I have not been able to get custom installations to work very well and definitely haven't gotten them to be persistent. So I'm not sure we have a workable solution. Here are some things that need to work:

- [ ] `conda install package` should work for the root environment for non-persistent installs.
- [ ] `pip install package` should generally work in the root environment too, it should use the root installed conda packages as a dependency base and drop this into the site-packages directory of the root conda install.
- [ ] `pip install --user package` should work. This should let users install packages into their home directories that also use the root conda installed packages as a dependency base.
- [ ] `conda create -n my-env package` should work and we should be able to open a notebook or console into that environment. A launcher icon should appear for this environment and if I click "new notebook" or "new console" the dropdown of optional kernels should show my custom environments.
- [ ] The custom conda envs in the bullet above should be able to be persistent too, i.e. install your environments into your home directory. Ideally these environments should be hardlinking to any already installed packages in the root environment so that this doesn't eat up too much diskspace in the user's home directory.
 - [ ] It would be nice if the user could install packages via apt, but I can see how that just may not be allowed because they could break the system in the docker file.
- [ ] We need to make sure that the install commands inside of R work too. Ideally they install R packages with conda, but I think the install.packages() command should work too.

**Comment by celine168** [comment]
> It would be nice if the user could install packages via apt, but I can see how that just may not be allowed because they could break the system in the docker file.

This would probably involve allowing `sudo` for users (if not, then we would have to use a different Linux OS instead of Ubuntu). There's a good comment/issue thread here about the topic: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/1021#issuecomment-438770079

**Comment by DarkBlaez** [comment]
I think we need to allow sudo for the user, what will happen is JupyterHub will see this as a recurring issue. No one is going to want to rebuild containers just to install dependencies for libraries. There are numerous use cases that need to be considered when moving to containers and Kubernetes. Experience with those platforms help here which would have guided the architecture to take this in to consideration. What work is being done to address this? simply allowing sudo would resolve as anyone running in Kubernetes will create a pv. Even the helm chart creates the pv so this would not be an issue

**Comment by rkevin-arch** [comment]
For python packages, if you install using `pip` you can specify a `--user` option which doesn't require root and installs it into the user's home directory (so it's synced across their pods anyway). I'm 99% sure conda has that ability as well (a preliminary google search says you can use `-p $HOME/local` or something to install to your home directory, and then just change the $PATH so you can run binaries from it). For apt packages, that will be more of an issue, but for very specific use cases we can just ask people to build it from source. TODO: research how feasible this is

**Comment by moorepants** [comment]
Yes, we've tried these things over the last year, things that would work in local installs, but have had lots of trouble. The goal here would be to provide a best practice for users to add new software to our hub. We should have the conda installs working that install to their home directory at a minmum, and then we can provide recommendations for other package managers if needed. The key thing is communicating how they can actually install to HOME.

**Comment by moorepants** [comment]
I recommend trying to setup some custom installs on a hub account to see what kinds of issues you hit.
